aten/src/ATen/core/ATenDispatch.h

```diff
@@ -73,34 +74,108 @@ class CAFFE2_API ATenDispatch {
+            .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA)
+        ));
+      } else {
+        add_c10_op_registration_(
+          torch::RegisterOperators().op(schema, torch::RegisterOperators::options()
+            .impl_unboxedOnlyKernel(c10::backendToTensorTypeId(backend), fn)
+            .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA)
+        ));
+      }
+    } else {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (op_tables_.find(schema) == op_tables_.end()) {
+        op_tables_.insert(std::make_pair(schema, ATenOpTable(schema)));
+      }
+      op_tables_.at(schema).registerOp(backend, reinterpret_cast<void*>(fn));
     }
-    op_tables_.at(schema).registerOp(backend, reinterpret_cast<void*>(fn));
     return *this;
   }
 
   template <class FuncType>
   ATenDispatch& registerVariableOp(const char* schema, FuncType* fn) {
-    std::lock_guard<std::mutex> lock(mutex_);
-    if (op_tables_.find(schema) == op_tables_.end()) {
-      op_tables_.insert(std::make_pair(schema, ATenOpTable(schema)));
+    if (op_was_already_moved_to_c10(schema)) {
+      add_c10_op_registration_(
+        torch::RegisterOperators().op(schema, torch::RegisterOperators::options()
+          .impl_unboxedAutogradKernel(fn)
+          .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA)
+      ));
+    } else {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (op_tables_.find(schema) == op_tables_.end()) {
+        op_tables_.insert(std::make_pair(schema, ATenOpTable(schema)));
+      }
+      op_tables_.at(schema).registerVariableOp(reinterpret_cast<void*>(fn));
     }
-    op_tables_.at(schema).registerVariableOp(reinterpret_cast<void*>(fn));
     return *this;
   }
 
   const ATenOpTable* getOpTable(const char* schema) const {
+    TORCH_CHECK(!op_was_already_moved_to_c10(schema), "Tried to get the globalAtenDispatch table for an op that is already moved to the c10 dispatcher");;
     auto iter = op_tables_.find(schema);
     TORCH_CHECK(iter != op_tables_.end(),
         "No functions are registered for schema ", schema);
     return &iter->second;
   }
 
  private:
+  static bool op_was_already_moved_to_c10(const char* schema) {
+    auto op_name = parse_operator_name(schema);
+    return aten_ops_already_moved_to_c10().count(op_name) != 0;
+  }
+
+  static c10::OperatorName parse_operator_name(const char* schema) {
+    // We can't depend on the jit function schema parser here, but parsing
+    // the op name is trivial. Let's just do it by hand.
+    std::string schema_str(schema);
+    size_t name_end_pos = schema_str.find_first_of(".(");
+    TORCH_CHECK(name_end_pos != std::string::npos, "Operator schema must contain a '(' character to start the argument list");
+    size_t overload_name_end_pos = name_end_pos + 1;
+    if (schema_str[name_end_pos] == '.') {
+      overload_name_end_pos = schema_str.find_first_of('(', name_end_pos);
+      TORCH_INTERNAL_ASSERT(schema_str[overload_name_end_pos] == '(');
```

Comment:
Shouldn't you assert that `overload_name_pos != std::string::npos`

Reply:
that would be a weaker assertion. The way it is currently asserts both, `!= npos` and that there is no off-by-one in the `find_first_of`.

Reply:
Assert `overload_name_pos != std::string::npos` in addition to the current assert. I don't see how your code currently asserts it `!= npos`. You just blindly index into the string using `overload_name_end_pos`.